pkg/controller/persistentvolume/persistentvolume_recycler_controller.go

```diff
@@ -94,16 +136,56 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
 	return recycler, nil
 }
 
+func (recycler *PersistentVolumeRecycler) shouldRecycle(pv *api.PersistentVolume) error {
+	if pv.Spec.ClaimRef == nil {
+		return fmt.Errorf("Volume does not have a reference to claim")
+	}
+	if pv.Status.Phase == api.VolumeReleased {
+		// This is the first attempt to recycle, allow.
+		return nil
+	}
+	if pv.Status.Phase != api.VolumeFailed {
+		return fmt.Errorf("The volume is not in a final state")
+	}
+
+	// The volume is Failed, should we retry recycling?
+	status, found := recycler.failedVolumes[pv.Name]
+	if !found {
+		// We don't know anything about this volume. The controller has been
+		// restarted or the volume has been marked as Failed by another
+		// controller. Recycle/delete this volume as if it was just Released.
+		glog.V(5).Infof("PersistentVolume[%s] not found in local cache, recycling", pv.Name)
+		return nil
+	}
+	glog.V(5).Infof("PersistentVolume[%s] %d attempts failed so far", pv.Name, status.retryCount)
+	if status.retryCount >= recycler.maximumRetryCount {
+		glog.V(5).Infof("PersistentVolume[%s] skipping recycling, %d attempts failed", pv.Name, status.retryCount)
+		return fmt.Errorf("%d attempts failed", status.retryCount)
+	}
+
+	// retry-count is fine, check the timestamp
+	expectedRetry := status.lastAttempt.Add(recycler.syncPeriod)
+	if time.Now().After(expectedRetry) {
+		glog.V(5).Infof("PersistentVolume[%s] retrying recycle after timeout", pv.Name)
+		return nil
+	}
+	// It's too early
+	glog.V(5).Infof("PersistentVolume[%s] skipping recycle, it's too early: now: %v, next retry: %v", pv.Name, time.Now(), expectedRetry)
+	return fmt.Errorf("Too early after previous failure")
+}
+
 func (recycler *PersistentVolumeRecycler) reclaimVolume(pv *api.PersistentVolume) error {
-	if pv.Status.Phase == api.VolumeReleased && pv.Spec.ClaimRef != nil {
+	glog.V(5).Infof("Recycler: checking PersistentVolume[%s]\n", pv.Name)
+	if err := recycler.shouldRecycle(pv); err == nil {
 		glog.V(5).Infof("Reclaiming PersistentVolume[%s]\n", pv.Name)
 
 		latest, err := recycler.client.GetPersistentVolume(pv.Name)
 		if err != nil {
 			return fmt.Errorf("Could not find PersistentVolume %s", pv.Name)
 		}
-		if latest.Status.Phase != api.VolumeReleased {
-			return fmt.Errorf("PersistentVolume[%s] phase is %s, expected %s.  Skipping.", pv.Name, latest.Status.Phase, api.VolumeReleased)
+
+		if err = recycler.shouldRecycle(latest); err != nil {
```

Comment:
You're right.   This line makes no sense, as it will never be true:

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go#L105

Unless I'm missing something, we don't need a 2nd check.  The 1st check can be broken in 2:

```
if phase != released, return error
if claimRef == nil, return error
...
the rest of the code...
```